### PR TITLE
Bytedance ImageX Adapter

### PR DESCRIPTION
**Pull request by ExerciseBook**
**Fixes #0000**

**Changes proposed in this pull request:**

To add the [ImageX](https://www.volcengine.com/product/imagex) adapter to fof-upload.

**Reviewers should focus on:**
- [ ] Is `composer.json` correct? Maybe I forget to add the suggested package into this file.
- [ ] Is my code works well without installing [exercisebook/flysystem-imagex](https://packagist.org/packages/exercisebook/flysystem-imagex) ? If well, it can not see ImageX options on admin panel.

**Screenshot**
Admin
Admin

Upload image file without setting template.
We can see that there is no template ID at the end of the URL.
Upload image file without setting template.

Upload image file with template setting.
We can see that there is a template ID at the end of the URL.
Upload image file with template setting.

Upload a non-image file but with template setting on.
Upload a non-image file but with template setting on.

```php
    protected function generateUrl(File $file)
    {
        $type = mb_strtolower($file->type);
        $path = $file->getAttribute('path');
        $template = $this->arrConfig["template"];

        if (Str::startsWith($type, "image/") && $template) {  // Check is image file or not.
            $url = '//' . $this->config->domain . '/' . $this->uriPrefix . '/' . $path . '~' . $template . '.image';
        } else {
            $url = '//' . $this->config->domain . '/' . $this->uriPrefix . '/' . $path;
        }
        $file->url = $url;
    }
```

**Confirmed**

- [ ] Frontend changes: tested on a local Flarum installation.
- [ ] Backend changes: tests are green (run `composer test`).

**Required changes:**

None

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -39,11 +39,13 @@
...
         "league/flysystem-aws-s3-v3": "Uploads to AWS S3 using API version 3.",
-        "overtrue/flysystem-qiniu": "Uploads to QiNiu using API."
+        "overtrue/flysystem-qiniu": "Uploads to QiNiu using API.",
+        "exercisebook/flysystem-imagex": "Uploads to ImageX using API."
     },
     "require-dev": {
         "league/flysystem-aws-s3-v3": "^1.0",
-        "overtrue/flysystem-qiniu": "1.0.4"
+        "overtrue/flysystem-qiniu": "1.0.4",
+        "exercisebook/flysystem-imagex": "<1.0"
```
Comment (TheSnowfield) [inline composer.json:48]: What's the "<1.0" meaning? I guess meant to require the package version that lower than 1.0? I have never seen this expression. 🤔

**Review by TheSnowfield — COMMENTED** [review]
Reviewed.

File: composer.json (hunk 1 above)
Comment (ExerciseBook) [reply composer.json:48]: <1.0 for flysystem V1
>=1.0 for flysystem V2

(Flarum uses flysystem V1)

**Comment by luceos** [comment]
Hi there and thank you for suggesting this upload adapter for fof/upload.

In order to keep this extension maintainable with the few number of people currently involved with it, we do not accept pull requests or feature requests for new upload adapters. The reason for this is that we would need to test the adapter on each upgrade of Flarum, plus whenever changes occur on this API of the upload service.

As a small team maintaining many extensions isn't always easy. Reducing the workload by making clear choices and what we do and don't support helps.

The best option you have is to (have someone) create a completely new extension that registers the additional upload adapter. The maintenance of that extension is then dispersed away from our team. The best way to add your adapter is by listening to the [Collecting event](https://github.com/FriendsOfFlarum/upload/blob/master/src/Events/Adapter/Collecting.php) using the `Flarum\Extend\Event` extender.

If you have any questions about the implementation, feel free to open a discussion on the [dev tag on discuss](https://discuss.flarum.org/t/dev).

Thank you for your understanding.

**Comment by ExerciseBook** [comment]
https://github.com/ExerciseBook/fof-upload-imagex 

A fof-upload extension was uploaded to packagist.

Thanks for your review @luceos .


This is a really late update for this pull request.

**Comment by luceos** [comment]
That's amazing @ExerciseBook 👍

**Pull request closed without merging**